### PR TITLE
#3273 Remove/Replace functionality

### LINKED ISSUE #3273: Temperature Sensor Replace/Remove buttons
## Is your feature request related to a problem? Please describe.

[In the designs for SensorEditPage](https://marvelapp.com/prototype/55aea8f/screen/75648030), there are buttons to "Replace" and "Remove" Sensors. 

## Describe the solution you'd like

TBD. Decide if we need both. Maybe "Replace":

1. Hit cog -> Sensor edit page
2. "Change sensor" button
3. Shows new sensor step 1 select*
4. return to edit page
5. usual input for everything there
6. Save all the details from old sensor (except macAddress and UUID)

*On selecting a sensor in step 3:

1. If its macAddress is the same as a `Sensor` record already on device, reject if `isActive` already (or do we just filter them out? Based on `isActive` or if currently related to a `Location`). If not `isActive`, make it active and associate with the `Location` (idk if we need to update movements records or anything, I'd think the relations of log -> location would already be enough).
2. If a new macAddress just add new `Sensor`, relate things.
3. old `Sensor.isActive` set to false, remove relation to `Location`.
4. If you fail to select one (i.e. back/cancel), keep the current configuration untouched, back to edit page.

I think some flavour of that behaviour ensures there is always one `Sensor` associated to a `Location`. For MVP do we need a way to set a Location/Sensor pair to `isActive` false? (We don't have Location.isActive in mobile yet.)

## Implementation

- Extract out the FlatList in `NewSensorStepOne` to reusable `SensorPicker`
  - Include the useEffect stuff
  - Add prop `onRowPress`, a call back that handles what to do when a sensor is selected from the list (passed through to ScanRow)
  - `onPress` prop for ScanRow...
  - Move redux dispatches to of `ScanRow` to `NewSensorStepOne`
  - `NewSensorStepOne` use `SensorPicker`
  - `NewSensorStepOne` dispatch methods in onRowPress of `SensorPicker`
- `EditSensorPage`, add button to "Replace" sensor as in designs
- Modal (or page? whatever is easier) that uses `SensorPicker`
- Currently the results in the list filter out sensors that are already in the DB according to `macAddress`. Change the filter to allow results that exist, but are currently `isActive == false`
- `OnRowPress` does the thing:
  - Make old sensor not isActive, remove relation to location
  - upsert selected sensor, isActive true, relate to the location. Assumes `name`, `location`, and `logInterval` of old sensor. (Note using upsert because the selected sensor may already have a record, but just not isActive)
  - redux `sensor` state is updated. New sensor is in there, old inactive sensor is not in there. `EditingId` updated with new sensor id
  - Adjust redux init/preload state to include inactive sensors. Don't display them in module page, don't attempt to sync them...
- Assignee definitely tests the following before PR:
  - New sensor works as normal
  - Edit fridge/sensor, replace sensor
    - SensorEditPage updates correctly and saves correctly
    - Navigating back to the list of fridges (Vacc module page) shows the updated sensor (i.e. macAddress should have changed) 

## Describe alternatives you've considered

Do nothing.

## Additional context

image

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3273 

## Change summary

- Adds replace sensor functionality
- Adds remove sensor functionality

## Testing

- [ ] When you remove a sensor, the sensor itself is set as isActive= false and removed from the vaccines page list. 
- [ ] Removed sensors are re-findable through scanning for a sensor
- [ ] When connecting/reenabling an old sensor, the old location and configs are reused, but updated with the details set in the new sensor flow
- [ ] When you replace a sensor, the old sensor is set as `isActive = false`, and the location is switched to the new sensor (with the temperature breach configs also)

### Related areas to think about

- I think there might be something here with "last sensor download" - do we want that to actually be last download from the location...? I am not sure!

### CODE REVIEW

File: src/widgets/AfterInteractions.js [hunk 1]
```diff
@@ -7,9 +7,14 @@ import { SUSSOL_ORANGE } from '../globalStyles/index';
...
   const [ready, setReady] = useState(false);
+  const handler = useRef();
 
   useEffect(() => {
-    InteractionManager.runAfterInteractions(() => setReady(true));
+    handler.current = InteractionManager.runAfterInteractions(() => {
+      setReady(true);
+      handler.current = null;
+    });
+    return () => handler.current?.cancel();
```
Comment (joshxg) [inline src/widgets/AfterInteractions.js:17]: stop doing stuff if unmounted, basically

File: src/widgets/PaperModal/PaperConfirmModal.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      <PageButton onPress={onCancel} text={cancelText} textStyle={{ textTransform: 'uppercase' }} />
+      <PageButton
+        onPress={onConfirm}
+        text={confirmText}
+        textStyle={{ textTransform: 'uppercase', color: WHITE }}
+        style={{ backgroundColor: SUSSOL_ORANGE }}
+      />
+    </FlexRow>
+  </FlexColumn>
+);
```
Comment (joshxg) [inline src/widgets/PaperModal/PaperConfirmModal.js:36]: somewhat reuseable. 

I'm sure it'll need changes when it is actually reused 🤣

File: src/widgets/SensorHeader/LastSensorDownload.js [hunk 3]
```diff
@@ -95,7 +95,7 @@ const stateToProps = (state, props) => {
   const isDownloading = selectIsDownloading(state, macAddress);
 
   const sensor = selectSensorByMac(state, macAddress);
-  const { logDelay } = sensor;
+  const { logDelay } = sensor ?? {};
```
Comment (joshxg) [inline src/widgets/SensorHeader/LastSensorDownload.js:98]: 🏢  🚗  🐛  office road bug?

File: src/widgets/PaperModal/PaperModalContainer.js [hunk 4]
```diff
@@ -19,7 +19,7 @@ export const PaperModalContainer = ({ isVisible, onClose, children }) => {
   return (
     <Modal
       visible={isVisible}
-      animationType="fades"
+      animationType="fade"
```
Comment (joshxg) [inline src/widgets/PaperModal/PaperModalContainer.js:22]: 🏖️  🐞  off road bug?

File: src/reducers/Entities/SensorReducer.js [hunk 5]
```diff
@@ -72,6 +73,18 @@ export const SensorReducer = (state = initialState(), action) => {
...
+      const { payload } = action;
+      const { macAddress, id } = payload;
+      const { byId, editingId } = state;
+
+      const oldSensor = byId[editingId];
+      const newSensor = { ...oldSensor, macAddress, id };
+      const newById = { ...byId, [id]: newSensor };
+
+      return { ...state, byId: newById, replacedId: editingId, editingId: id };
+    }
```
Comment (joshxg) [inline src/reducers/Entities/SensorReducer.js:86]: this was actually a kinda cool thing to just switch around the ids and it seemed to work (HOPEFULLY 🤞 )

File: src/selectors/Entities/sensor.js [hunk 6]
```diff
@@ -9,12 +9,23 @@ export const selectSensorsById = state => {
...
+export const selectSensorByMac = (state, mac) => {
+  const sensors = selectSensors(state);
+  const foundSensor = sensors.find(({ macAddress }) => macAddress === mac);
   return foundSensor;
 };
 
+export const selectActiveSensors = state => {
+  const sensors = selectSensors(state);
+  const filtered = sensors.filter(({ isActive }) => isActive);
+  return filtered;
```
Comment (joshxg) [inline src/selectors/Entities/sensor.js:26]: non `isActive` sensors will now be floating around and need to be filtered out

**Comment by joshxg** [comment]
I used this A LOT to test this as I found it to be the easiest to ensure the persisted changes were accurate:

`adb shell "run-as com.msupplymobile cat /data/data/com.msupplymobile/files/default.realm" > ~/mobile.realm`

**Comment by mark-prins** [comment]
I'd rather like a `CANCEL` button on the replace modal. Anyone else feel the same?

File: src/widgets/AfterInteractions.js (hunk 1 above)
Comment (mark-prins) [reply src/widgets/AfterInteractions.js:17]: looks fine - but on my tablet the edit sensor page just spins endlessly... reverting these changes makes it work

File: src/widgets/PaperModal/PaperConfirmModal.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      {Icon}
+      <Spacer space={30} vertical />
+      <Text style={localStyles.questionText}>{questionText}</Text>
+    </FlexView>
+    <FlexRow justifyContent="space-evenly" alignItems="center" style={{ width: '100%' }} flex={1}>
+      <PageButton onPress={onCancel} text={cancelText} textStyle={{ textTransform: 'uppercase' }} />
+      <PageButton
+        onPress={onConfirm}
+        text={confirmText}
+        textStyle={{ textTransform: 'uppercase', color: WHITE }}
```
Comment (mark-prins) [inline src/widgets/PaperModal/PaperConfirmModal.js:31]: these should probably be in `localStyles`

File: src/widgets/SensorHeader/LastSensorDownload.js (hunk 3 above)
Comment (mark-prins) [reply src/widgets/SensorHeader/LastSensorDownload.js:98]: ah nice. I did the same in another PR too!

**Review by mark-prins — COMMENTED** [review]
Golly, that's a lot of code to read through!
Looks awesome, and I like how it's working. One question around the `AfterInteractions` which I'm playing with now..

File: src/widgets/AfterInteractions.js (hunk 1 above)
Comment (mark-prins) [reply src/widgets/AfterInteractions.js:17]: 🤷‍♂️  can't get it to either work or fail reliably. The fridge display is also spinning most of the time, so there's something amiss with the interactions, just not sure what.

**Review by mark-prins — APPROVED** [review]
I've got a bit of an issue with the tablet. Emulator is happy, may have to make another issue if someone else reproduces the problem.

**Comment by joshxg** [comment]
> Golly, that's a lot of code to read through!

Yeah, sorry. Shouldv'e split into a replace and remove PRs

> looks fine - but on my tablet the edit sensor page just spins endlessly

Hm. I thought I had fixed it - it was happening on an emulator for me. The weird thing being that ` `VaccinePage` was fine, but the edit page wasn't. The addition of `DURATION_OFFSET` in the hook the flashing circle uses helped it for me, but I think I'll just revert the change you've pointed out and think about it some more. 

Weird! I like that `AfterInteractions` component ! hehe

**Comment by joshxg** [comment]
With the infinite spinning:
ec78f5a
48e2261
171643d

- Seems the animation being used for the `circle` was registering a handler on the interaction manager (when `iteractionCount !== 'infinite'`. The interaction manager is then queueing up functions in `runAfterInteractions` until all handlers are cleared. 
- Then, I think the function registered in `setInterval` was being called before the `runAfterInteractions` queued tasks, registering new interactions on the manager before it properly registered as 'finished'.

Can test this theory by increasing the duration between the animation of the circle and the duration between each set interval.

Anyway, I removed this way of animating and just used `iterationCount="infinite"`

File: src/actions/Bluetooth/SensorScanActions.js [hunk 8]
```diff
@@ -19,6 +19,16 @@ const scanStart = () => ({ type: SCAN_ACTIONS.SCAN_START });
 const scanStop = () => ({ type: SCAN_ACTIONS.SCAN_STOP });
 const sensorFound = macAddress => ({ type: SCAN_ACTIONS.SENSOR_FOUND, payload: { macAddress } });
 
+const isNewlyFoundSensor = (macAddress, alreadyScannedMacAddresses) => {
+  const alreadyFound = alreadyScannedMacAddresses?.includes(macAddress);
+  const alreadySaved = UIDatabase.get('Sensor', macAddress, 'macAddress');
+  const isActive = alreadySaved?.isActive;
+  const isAlreadySavedAndActive = alreadySaved && isActive;
+
+  // A newly found sensor are sensors which are not active in the database, and not found already
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/actions/Bluetooth/SensorScanActions.js]: 82f63f9

File: src/widgets/PaperModal/PaperConfirmModal.js (hunk 7 above)
Comment (joshxg) [reply src/widgets/PaperModal/PaperConfirmModal.js]: 235d907

**Comment by mark-prins** [comment]
The reverting of the animation interaction handling has worked for me. Perhaps we raise an issue to play with that, and merge in this change as it is now? I really like what you did with the interactions & animations, would be great to have it back :)

**Comment by joshxg** [comment]
Happy to merge, just didn't want to merge it in and break everyones branches! It's working well for me now, so .. 🤞

**Pull request merged**